File: Programmers/체육복.js
// https://programmers.co.kr/learn/courses/30/lessons/42862

function solution(n, lost, reserve) {
  let realReserve = reserve.filter((v) => !lost.includes(v));
  let realLost = lost.filter((v) => !reserve.includes(v));
  let answer = n - realLost.length;

  realLost.map((n) => {
    if (realReserve.includes(n - 1)) {
      answer++;
      realReserve = realReserve.filter((v) => v !== n - 1);
      return;
    }
    if (realReserve.includes(n + 1)) {
      answer++;
      realReserve = realReserve.filter((v) => v !== n + 1);
      return;
    }
  });

  return answer;
}

// const n = 5;
// const lost = [2, 4];
// const reserve = [1, 3, 5];

// const n = 5;
// const lost = [2, 4];
// const reserve = [3];

const n = 3;
const lost = [3];
const reserve = [1];

console.log(solution(n, lost, reserve));

File: Programmers/체육복.swift
//https://programmers.co.kr/learn/courses/30/lessons/42862?language=swift
//체육복

import Foundation

func solution(_ n:Int, _ lost:[Int], _ reserve:[Int]) -> Int {
    var answer: Int = 0
    var realReserve: [Int] = reserve.filter { !lost.contains($0) }
    let realLost: [Int] = lost.filter { !reserve.contains($0) }
    
    for i in 1...n {
        if !realLost.contains(i) {
            answer += 1
            continue
        }
        
        if let smallSizeIndex = realReserve.firstIndex(of: i - 1) {
            answer += 1
            realReserve.remove(at: smallSizeIndex)
            continue
        }
        
        if let largeSizeIndex = realReserve.firstIndex(of: i + 1) {
            answer += 1
            realReserve.remove(at: largeSizeIndex)
            continue
        }
    }
    
    return answer
}

        
//let n: Int = 5
//let lost: [Int] = [2, 4]
//let reserve: [Int] = [1, 3, 5]
////5

let n: Int = 5
let lost: [Int] = [2, 4]
let reserve: [Int] = [3]
//4

//let n: Int = 3
//let lost: [Int] = [3]
//let reserve: [Int] = [1]
////2

print(solution(n, lost, reserve))
